dodo.py

```diff
@@ -0,0 +1,400 @@
+            actions=[
+                U.do(
+                    "echo",
+                    f"::group::Build {P.OWNER}/{image}- using system's architecture",
+                ),
+                U.do(
+                    "docker",
+                    "buildx",
+                    "build",
+                    *["-t" + tag for tag in image_tags],
+                    "-f",
+                    dockerfile,
+                    "--build-arg",
+                    "OWNER=" + P.OWNER,
+                    str(image_dir),
+                    "--load",
+                ),
+            ],
+            file_dep=[dockerfile],
+            uptodate=[False],
+        )
+
+        yield dict(
+            name=f"build_summary:{image}",
+            doc="Brief summary of the image built - defaulting to using the latest tag",
+            actions=[
+                ["echo", "\n \n ⚡️ Build complete, image size:"],
+                U.do("docker", "images", image_tags[0], "--format", "{{.Size}}"),
+                U.do("echo", "::endgroup::"),
+            ],
+        )
+
+
+def task_docker_save_images():
+    """Save the built Docker images - these will be stored as CI artifacts.
+    This is needed to pass images across jobs in GitHub Actions as each job runs in a separate container."""
+
+    if U.IS_CI:
+        images_ids = U.get_images()
+
+        return dict(
+            targets=[U.CI_IMAGE_TAR],
+            actions=[
+                U.do("echo", f"Saving images to: {P.CI_IMG}"),
+                U.do("mkdir", "-p", P.CI_IMG),
+                U.do(
+                    "docker",
+                    "save",
+                    *images_ids,
+                    "-o",
+                    U.CI_IMAGE_TAR,
+                ),
+            ],
+        )
+
+
+def task_docker_test():
+    """Test Docker images - needs to be run after `docker_build`"""
+
+    if U.IS_CI & Path(U.CI_IMAGE_TAR).exists():
+        """Since we are running in a CI environment and within a separate job than the one where the images are built,
+        we need to load the images from the CI_IMAGE_TAR"""
+
+        yield dict(
+            name="load_images",
+            doc="Load and inspect Docker images",
+            actions=[
+                U.do("docker", "load", "--input", U.CI_IMAGE_TAR),
+                (U.inspect_image, [U.get_images()[-1]]),
+                U.do("docker", "images"),
+            ],
+        )
+
+    for image in P.ALL_IMAGES:
+
+        test_dirs = get_test_dirs(image)
+
+        yield dict(
+            name=f"test:{image}",
+            doc="Run the test suite for the images - will always run the tests if an image is built",
+            uptodate=[False],
+            actions=[
+                (U.set_env, [f"{P.OWNER}/{image}"]),
+                (
+                    U.do(
+                        *U.PYTEST_ARGS,
+                        *test_dirs,
+                    )
+                ),
+            ],
```

Comment:
That was my plan at the beginning - though since [`run_tests.py` uses plumbum ](https://github.com/jupyter/docker-stacks/blob/2e7b2562e0dd5778fbfce67d757b707086118638/tests/run_tests.py?rgh-link-date=2022-03-04T14%3A01%3A38Z#L20) these two (doit task and run_test.py) end up being incompatible. 

To avoid duplication the easier way would be to run tests directly through doit tasks and get rid of `run_tests.py` altogether

Reply:
It would be nice to know the reason for this conflict.
I would like to continue to use plumbum :)

Reply:
I think I made it work with run_tests. And now there is no duplication.

Also, I would like to have a way to test only one image locally.
Right now I use run_tests.py for that.